File: sortingAlgorithms/day17/heapSort.py
def heapify(arr, n, i):
    largest = i
    left = 2*i+1
    right = 2*i+2
    if left < n and arr[i] <arr[left]:
        largest = left

    if right < n and arr[largest]<arr[right]:
        largest = right
    
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def buildHeap(arr, n):
    for i in range(n//2-1, -1, -1):
        heapify(arr, n, i)

def heapSort(arr):
    n = len(arr)
    buildHeap(arr, n)

    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)

arr = [4, 10, 3, 5, 1]
heapSort(arr)
print("Sorted array:", arr)
# 输出: Sorted array: [1, 3, 4, 5, 10]

File: sortingAlgorithms/day19/heapSort.cpp
#include <iostream>
#include <vector>
using namespace std;
/*
在 heapSort 中，首先需要 O(n) 的时间来构建堆，然后每次从堆中移除最大元素，
并且每次都要调用 heapify 来恢复堆的性质。移除每个元素的时间复杂度为 O(log n)，
共需要执行 n 次移除操作，因此 heapSort 的时间复杂度为 O(n log n)。
*/

void heapify(vector<int> &arr, int n, int root)
{
    int largest = root;
    int left = largest * 2 + 1;
    int right = largest * 2 + 2;

    if (left < n && arr[left] > arr[largest])
        largest = left;

    if (right < n && arr[right] > arr[largest])
        largest = right;

    if (root != largest)
    {
        swap(arr[root], arr[largest]);
        heapify(arr, n, largest); // 递归遍历子树，确保修改后都符合大根堆的性质
    }

    return;
}

void buildHeap(vector<int> &arr, int n)
{
    for (int i = n / 2 - 1; i >= 0; i--)
    {
        heapify(arr, n, i);
    }
}

void heapSort(vector<int> &arr)
{
    int n = arr.size();
    buildHeap(arr, n);

    for (int i = n / 2 - 1; i >= 0; i--)
    {
        swap(arr[0], arr[i]); // 将堆顶元素（最大值）放到数组末尾
        heapify(arr, i, 0);
    }
}
